File: src/toolchains/pip.py
import subprocess
import json
import shutil


def scan():
    result = {
        "tool": "pip",
        "available": False,
        "python_version": None,
        "outdated": [],
        "issues": [],
        "status": "ok",
    }

    if not shutil.which("python3"):
        result["status"] = "skipped"
        result["issues"].append("Python 3 not installed")
        return result

    try:
        py_v = subprocess.run(
            ["python3", "--version"], capture_output=True, text=True, timeout=5
        )
        result["python_version"] = py_v.stdout.strip()

        if not shutil.which("pip3"):
            return result

        outdated = subprocess.run(
            ["pip3", "list", "--outdated", "--format=json"],
            capture_output=True,
            text=True,
            timeout=30,
        )
        if outdated.returncode == 0 and outdated.stdout.strip():
            data = json.loads(outdated.stdout)
            result["outdated"] = [
                {
                    "name": pkg["name"],
                    "current": pkg.get("version", "?"),
                    "latest": pkg.get("latest_version", "?"),
                }
                for pkg in data
            ]

        result["status"] = "ok" if not result["outdated"] else "warning"
        if result["outdated"]:
            result["issues"].append(
                f"{len(result['outdated'])} outdated pip package(s)"
            )
    except Exception as e:
        result["status"] = "error"
        result["issues"].append(str(e))

    return result

File: src/toolchains/pip.rs
use super::*;

pub async fn scan() -> ScanResult {
    if !which("python3") {
        return ScanResult::skipped("Python not installed");
    }

    let mut result = ScanResult::new("pip");

    let has_pip3 = which("pip3");

    if has_pip3 {
        let (py_ver, pip_ver, outdated) = tokio::join!(
            run_cmd("python3", &["--version"], None),
            run_cmd("pip3", &["--version"], None),
            run_cmd("pip3", &["list", "--outdated", "--format=json"], None)
        );

        if let Ok(ver) = py_ver {
            result.python_version = Some(ver);
        }
        if let Ok(ver) = pip_ver {
            result.version = Some(ver);
        }
        if let Ok(out) = outdated {
            if let Ok(packages) = serde_json::from_str::<Vec<serde_json::Value>>(&out) {
                for pkg in packages {
                    result.outdated.push(PackageInfo {
                        name: pkg["name"].as_str().unwrap_or("?").to_string(),
                        current: pkg["version"].as_str().unwrap_or("?").to_string(),
                        latest: pkg["latest_version"].as_str().unwrap_or("?").to_string(),
                    });
                }
            }
        }
    } else {
        if let Ok(ver) = run_cmd("python3", &["--version"], None).await {
            result.python_version = Some(ver);
        }
    }

    result.status = if result.outdated.is_empty() {
        "ok".into()
    } else {
        "warning".into()
    };

    result
}
